draft blog post for writing a toy optimizer (#63)
pypy/pypy.org: posts/2022/07/toy-optimizer.rst

.. title: Implementing a Toy Optimizer
.. slug: toy-optimizer
.. date: 2022-07-22 15:00:00 UTC
.. tags:
.. category:
.. link:
.. description:
.. type: rest
.. author: Carl Friedrich Bolz-Tereick

In this blog post I want to show the complete code (in Python3) of how a very
simple optimizer for sequences of operations can work. These algorithms could

> be part of a (really simple) compiler, or a JIT.

Reviewer: Could you add another sentence about motivation: is this a tutorial on what the PyPy JIT does and at what stage, is this a general JIT theory tutorial, is this the first in a series of "How JITs works", ?

Reviewer: This is touched on near the end
> And indeed, what I describe in this post is very similar to the one
PyPy's JIT optimizer uses

so maybe rephrase it here:
> and in fact the optimizer style we end up with is very similar to the one in PyPy's JIT

Author: thanks, did this

computing it once. In this post I want to show an optimizer that can do this
(and some related) optimizations.

Looking at the IR we notice that the input expression has been linearized
into a sequence of operations, and all the intermedia results have been given
unique variable names. The value that every variable is assigned is computed
by the right hand side, which is some operation consisting of an operand and an
arbitrary number of arguments. The arguments of an operation are either
themselves variables or constants.

I will not at all talk about the process of translating the input program
into the IR. Instead, I will assume we have some component that does this
translation already. The tests in this blog post will construct small
snippets of IR by hand. I also won't talk about what happens after the
optimization (usually the optimized IR is translated into machine code).

Implementing the Intermediate Representation
=============================================

Let's start modelling the intermediate representation with Python classes.
First we define a base class of all values that can be used as arguments in

> operations, and let's also add a class that represents constants:

Reviewer: ```suggestion
operations, and let's also add a dataclass_ that represents constants:
.. _dataclass: https://realpython.com/python-data-classes/
```

Author: removed the data classes entirely

.. code:: python

    import pytest
    from dataclasses import dataclass
    from typing import Optional, Any

    class Value:
        pass

    @dataclass(eq=False)
    class Constant(Value):
        value : Any

One consequence of the fact that every variable is assigned to only once is
that variables are in a one-to-one correspondence with the right-hand-side of
their unique assignments. That means that we don't need a class that represents
variables at all. Instead, it's sufficient to have a class that represents an

> operation, and that is the same as the variable that it defines:

Reviewer: ```suggestion
operation (the right-hand side), and that by definition is the same as the variable (left-hand side) that it defines:
```

                               args=[Constant(value=1)]),
                     Constant(value=17)]),
     Operation(name='add',
               args=[Operation(name='mul',
                               args=[Operation(name='getarg',
                                               args=[Constant(value=0)]),
                                     Operation(name='add',
                                               args=[Operation(name='getarg',
                                                               args=[Constant(value=1)]),
                                                     Constant(value=17)])]),
                     Operation(name='add',
                               args=[Operation(name='getarg',
                                               args=[Constant(value=1)]),
                                     Constant(value=17)])])]

It's impossible to see what is going on here, because the `Operations` in the
basic block appear several times, once as elements of the list but then also as
arguments to operations further down in the list. So we need some code that
turns things back into a readable textual representation, so we have a chance
to debug.

.. code:: python

>     def bb_to_str(l : Block, varprefix : str = "var"):

Reviewer: `l` is not the best choice for an argument name, it took me quite a while to figure out where it was used when scanning the code
```suggestion
    def bb_to_str(l : Block, varprefix : str = "var"):
```

Author: I renamed the argument to bb like everywhere else

define sequences of operations and print them in a readable way. Next we need a
central data structure that is used when actually optimizing basic blocks.

Storing Equivalences between Operations Using a Union-Find Data Structure
=========================================================================

When optimizing a sequence of operations, we want to make it less costly to
execute. For that we typically want to remove operations (and sometimes
replace operations with less expensive ones). We can remove operations if
they do redundant computation, like case of the duplicate `add(v1, 17)` in
the example. So what we want to do is to turn the running input sequence::

    v0 = getarg(0)
    v1 = getarg(1)
    v2 = add(v1, 17)
    v3 = mul(v0, v2)
    v4 = add(v1, 17)
    v5 = add(v3, v4)

Into the following optimized output sequence::

    optvar0 = getarg(0)
    optvar1 = getarg(1)
    optvar2 = add(optvar1, 17)
    optvar3 = mul(optvar0, optvar2)
    optvar4 = add(optvar3, optvar2)

We left out the second ``add`` (which defines ``v4``), and then replaced the
usage of ``v4`` with ``v2`` in the final operation.

> 

Reviewer: maybe add line numbers, and then you can say "We left out line ... and replaced v4 with v2 in line ...

Author: yeah, that would be great, unfortunately the style of numbered blocks is really different than regular ones :-(

![image](https://user-images.githubusercontent.com/85942/179674543-c3802fc6-35af-4e4e-b001-d4f1b06a7a9d.png)

so I'm inclined to leave it as it is.

point of view, sticking everything into a single function like in `optimize`
above is obviously not great, and if you wanted to do this for real you would
try to split the cases into different functions that are individually
digestible, or even use a DSL that makes the pattern matching much more
readable. But the advantage of the architecture is that it's quite efficient,
it makes it possible to pack a lot of good optimizations into a single pass
over a basic block.

Of course this works even better if you are in a tracing context, where
everything is put into a trace, which is basically one incredibly long basic
block. And indeed, what I describe in this post is very similar to the one
PyPy's JIT optimizer uses. In a JIT context it's also quite important that the
optimizer itself runs quickly.

Various other optimizations are possible in this model. I plan to write a
follow-up post that show how to implement what is arguably PyPy's `most
important optimization`_.

.. _`most important optimization`: https://www.pypy.org/posts/2010/09/escape-analysis-in-pypys-jit-1780048403046080197.html

Some Further Pointers
======================

> This is a blog post, not a paper, but I still want to give some pointers to
> literature that can be read to understand the concepts that are introduced here
> in much bigger generality.

Reviewer: No need to demean this blog post, summarizing is sometimes harder than writing a paper.
```suggestion
This is a short introduction to many fundamental concepts that are
covered in more general literature.
```

Author: makes sense, wrote something along these lines
